archive/overlap.go

```diff
@@ -198,3 +209,199 @@ func mergeLargestChunkSpanInfos(spans []SpanInfo, order zbuf.Order) []SpanInfo {
+		}
+	}
+	return mc
+}
+
+type compactWriter struct {
+	ark   *Archive
+	ctx   context.Context
+	masks []ksuid.KSUID
+	tsd   tsDir
+	w     *chunkWriter
+}
+
+func (cw *compactWriter) Write(rec *zng.Record) error {
+	if cw.w != nil {
+		pos, firstTs, lastTs := cw.w.position()
+		if pos > cw.ark.LogSizeThreshold && lastTs != rec.Ts() {
+			// If we need to create a new chunk writer, we must ensure that the
+			// span for the current chunk leaves no gap between its last timestamp
+			// and the first timestamp of our next chunk, to ensure these chunks
+			// are seen as covering the entire span of the source chunks. Hence
+			// the use of 'chunkLastTs', and the lastTs check above to ensure we are
+			// not in a run of records with the same timestamp.
+			chunkLastTs := prevTs(rec.Ts(), cw.ark.DataOrder)
+			if _, err := cw.w.closeWithTs(cw.ctx, firstTs, chunkLastTs); err != nil {
+				return err
+			}
+			cw.w = nil
+		}
+	}
+	if cw.w == nil {
+		var err error
+		cw.w, err = newChunkWriter(cw.ctx, cw.ark, cw.tsd, FileKindDataCompacted, cw.masks)
+		if err != nil {
+			return err
+		}
+	}
+	return cw.w.Write(rec)
+}
+
+func (cw *compactWriter) abort() {
+	if cw.w != nil {
+		cw.w.abort()
+		cw.w = nil
+	}
+}
+
+func (cw *compactWriter) close(lastTs nano.Ts) error {
+	if cw.w == nil {
+		return nil
+	}
+	_, firstTs, _ := cw.w.position()
+	if _, err := cw.w.closeWithTs(cw.ctx, firstTs, lastTs); err != nil {
+		return err
+	}
+	cw.w = nil
+	return nil
+}
+
+func compactOverlaps(ctx context.Context, ark *Archive, s SpanInfo) error {
+	if len(s.Chunks) == 1 {
+		return nil
+	}
+	ss, err := newSpanScanner(ctx, ark, resolver.NewContext(), nil, nil, s)
+	if err != nil {
+		return err
+	}
+	defer ss.Close()
+	tsd := newTsDir(s.Span.Ts)
+	var masks []ksuid.KSUID
+	for _, c := range s.Chunks {
+		masks = append(masks, c.Id)
+	}
+	mw := &compactWriter{
+		ark:   ark,
+		ctx:   ctx,
+		masks: masks,
+		tsd:   tsd,
+	}
+	_, slast := spanToFirstLast(ark.DataOrder, s.Span)
+	if err := zbuf.CopyWithContext(ctx, mw, zbuf.PullerReader(ss)); err != nil {
+		mw.abort()
+		return err
+	}
+	return mw.close(slast)
+}
+
+// mergeCommonChunkSpans generates a SpanInfo slice by merging runs of
+// SpanInfos where a SpanInfo has some chunk in common with its previous
+// SpanInfo.
+func mergeCommonChunkSpans(spans []SpanInfo, order zbuf.Order) []SpanInfo {
+	if len(spans) < 2 {
+		return spans
+	}
+	var res []SpanInfo
+	run := []SpanInfo{spans[0]}
+outer:
+	for _, s := range spans[1:] {
+		for _, c := range s.Chunks {
+			if spanInfoContainsChunk(run[len(run)-1], c.Id) {
+				run = append(run, s)
+				continue outer
+			}
+		}
+		res = append(res, mergeSpanInfos(run, order))
+		run = []SpanInfo{s}
+	}
+	return append(res, mergeSpanInfos(run, order))
+}
+
+func Compact(ctx context.Context, ark *Archive) error {
+	return tsDirVisit(ctx, ark, nano.MaxSpan, func(_ tsDir, chunks []Chunk) error {
+		spans := alignChunksToSpans(chunks, ark.DataOrder, nano.MaxSpan)
```

Comment:
Why do we need this `alignChunksToSpans` call? Why not just do:
```golang
spanInfo := SpanInfo{tsDir.Span, chunks}
```

Reply:
If we did that, then we would always compact all of the chunks in a tsdir, not just those that shared some overlapping time span.